### PR TITLE
Code Editor options and context menus

### LINKED ISSUE #24812: Workbench editor: right click context menu with Open File Location
### Expected behavior
Right clicking an editor should show a context menu. One of the options should be `Open File Location`.

### Actual behavior
Context menu doesn't exist

### LINKED ISSUE #24840: Workbench: Add toolbar to editor, and move Editor options to right side of tab widget
Consider a toolbar on the script editor, a bit like PyCharm. It would contain buttons for execution (Start/Stop[Abort]).

Add Editor Options button on the right side of the QTabWidget, which will contain what the current `Editor` menu item contains

### PR DESCRIPTION
**Pull request by dtasev**
Note: This was rebased on #24790, to reuse some testing classes implemented there, so it currently shows a lot more files than have actually been changed. I'll merge that PR here when it is merged into master to remove that

**Description of work.**
Adds the following:
- Context menu (right click) on each editor tab. Currently only has `Show in Explorer`, more options can be added on request. Maybe `Close Others`, `Close All to the Left`, `Close All to the Right`
  - Try `Show in Explorer` on tabs saved to a file (should open the dir containing the file). Tested on Windows and Ubuntu. Errors are logged at `DEBUG` log level.
  - Try `Show in Explorer` on tabs not saved anywhere, should log at Notice that tab isn't saved (check if message makes sense)
  - [x] Try `Show in Explorer` on MAC. **It might log an error (which will be fixed in this PR) but shouldn't crash!**
- Middle-mouse click on a tab closes that tab. Middle mouse click on the tab bar (_not_ on a tab) does nothing
- Options Menu is now in the TabBar. All Actions should still work properly.
- Run/Abort buttons have been added. They should work the same as Options > Run/Abort

**To test:**
Click all the buttons in all the places for all the tabs.

Fixes #24812 
Fixes #24840 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqt/widgets/codeeditor/tab_widget/codeeditor_tab_presenter.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+from mantid.kernel import logger
+
+
+class ShowInExplorer(object):
+    @staticmethod
+    def open_directory(path):
+        if sys.platform == "win32":
+            os.startfile(os.path.dirname(path))
+        elif sys.platform == 'darwin':
+            subprocess.check_call(['open', '--', path])
```
Comment (DTasev) [inline qt/python/mantidqt/widgets/codeeditor/tab_widget/codeeditor_tab_presenter.py]: This needs to be verified that it doesn't error on Mac

File: qt/python/mantidqt/widgets/codeeditor/tab_widget/codeeditor_tab_view.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        container_widget = QWidget(self)
+        layout = QHBoxLayout(container_widget)
+        layout.setContentsMargins(0, 0, 0, 0)
+        layout.setSpacing(0)
+        container_widget.setLayout(layout)
+        self.setCornerWidget(container_widget, Qt.TopRightCorner)
+
+        run_button = QPushButton(container_widget)
+        run_button.setObjectName(self.RUN_BUTTON_OBJECT_NAME)
+        run_button.setIcon(get_icon("fa.play", color=QColor(73, 156, 84)))
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/codeeditor/tab_widget/codeeditor_tab_view.py]: Maybe extract the QColor as a named class-level variable with `GREEN` in the name?

File: qt/python/mantidqt/widgets/codeeditor/tab_widget/codeeditor_tab_view.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        run_button.clicked.connect(parent.execute_current)
+        layout.addWidget(run_button)
+
+        abort_button = QPushButton(container_widget)
+        abort_button.setObjectName(self.ABORT_BUTTON_OBJECT_NAME)
+        abort_button.setIcon(get_icon("fa.square", color=QColor(230, 84, 80)))
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/codeeditor/tab_widget/codeeditor_tab_view.py]: Similarly extract a named variable with `RED` in the name?

**Review by martyngigg — CHANGES_REQUESTED** [review]
I like the layout. I think it's much clearer with things being closer to the editor where they will be used.

A couple of things in addition to the in-line comments:
  - It's looks like there is a top-level `__init__.py` that has been added in error?
  - It also seems that "Ctrl+S" save shortcut has broken as I can't save scripts with it anymore?

File: qt/python/mantidqt/widgets/codeeditor/tab_widget/codeeditor_tab_presenter.py (hunk 1 above)
Comment (martyngigg) [reply qt/python/mantidqt/widgets/codeeditor/tab_widget/codeeditor_tab_presenter.py]: Confirmed that this does work on macOS. :tada:

**Comment by DTasev** [comment]
The latest commit removes the prompt when clicking File > Save Script or CTRL + S, as the file should just be saved

File: qt/python/mantidqt/widgets/codeeditor/interpreter.py [hunk 4]
```diff
@@ -46,14 +46,22 @@ def ask_for_filename(self):
             filename = None
         return filename
 
-    def save_if_required(self, confirm=True):
-        """Asks the user if the contents should be saved.
-
-        :param confirm: If True then show a confirmation dialog first to check we should save
+    def save_if_required(self, prompt_for_confirm=True, force_save=False):
+        """
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/codeeditor/interpreter.py]: Good docs :+1:

**Review by martyngigg — DISMISSED** [review]
This looks good now. It will make the editor easier to use in the case it is detached from the main window.

**Comment by SimonHeybrock** [comment]
There is a merge conflict now.

**Review by martyngigg — APPROVED** [review]
Confirmed that this still works after the merge.

**Comment by martyngigg** [comment]
The `OSX` failure is unrelated.

**Pull request merged**